Enhancement: Scarecrow song skip

Dumbledork01 opened the pull request:
**Enhancements**

* Pull out ocarina to summon Pierre rather than having to teach the Scarecrows the song first and then play the song later. (This is in many randomizers, so I consider it more of a time saver than a cheat.)

**Demonstration Videos**
Scarecrow Song Time Saver: https://youtu.be/_i-STqVWu5I

Dumbledork01 committed:
Added Skip Scarecrow Song

Added in the option for if the user has the time saver for Scarecrow song enabled and has their Ocarina pulled out in range of Pierre.
Added Skip Scarecrow Song
New enhancement under time-savers that allows the user to only pull out their ocarina to summon Pierre.

aMannus commented:
LGTM 👍 

Oh, actually. Could you add this to the new enhancement presets as well? So add an entry to the default preset setting it to 0, and add it to whatever preset you think fits best. I'd personally wouldn't add it to any other preset because it could be seen as cheaty, but I'll leave that up to you.

Dumbledork01 committed:
Added Default for Scarecrow summon

The scarecrow summon option now defaults to 0 and can be turned on individually. I haven't included it in any other presets so people don't feel like they're cheating.

briaguya0 commented:
as far as i can tell from looking through the code, free scarecrow isn't accounted for in the 3drando logic at all. that's good news for this PR (don't need to worry about updating anything in rando for it right now), but it does feel odd to me when considering entrance rando (logically you'd expect all scarecrow checks to be locked behind lake hylia access, but with free scarecrow that's not required)

Dumbledork01 commented:
@briaguya-ai Yeah that makes sense. Usually randomizers have a free-scarecrow option to be included with the logic, so I assume that would have to be selected for generating the randomizer to remove those checks. If that was the case, then if someone selected this enhancement with free scarecrow not being included in the logic of the seed they generated, then it would be more of a cheat I suppose.

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Scarecrow-Song-Skip

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Scarecrow-Song-Skip

Dumbledork01 committed:
Checks if scarecrow song is in possession

briaguya0 commented on soh/src/overlays/actors/ovl_En_Kakasi2/z_en_kakasi2.c:
```diff
@@ -118,9 +118,10 @@ void func_80A90264(EnKakasi2* this, GlobalContext* globalCtx) {
-        (fabsf(player->actor.world.pos.y - this->actor.world.pos.y) < this->maxSpawnDistance.y)) {
-
+    if (((BREG(1) != 0) && (this->actor.xzDistToPlayer < this->maxSpawnDistance.x) &&
+            (fabsf(player->actor.world.pos.y - this->actor.world.pos.y) < this->maxSpawnDistance.y)) ||
```
did a quick little reformat to try to better see the logic in this if statement (not recommending this as a code change)
```c

if (
    (
     (BREG(1) != 0) &&
     (this->actor.xzDistToPlayer < this->maxSpawnDistance.x) &&
     (fabsf(player->actor.world.pos.y - this->actor.world.pos.y) < this->maxSpawnDistance.y)
    ) ||
    (
     (CVar_GetS32("gSkipScarecrow",0) != 0) &&
     (globalCtx->msgCtx.ocarinaAction == OCARINA_ACTION_FREE_PLAY) &&
     gSaveContext.scarecrowSpawnSongSet
    )
   )
```
it seems like the logic here is skipping all the distance/position checks when skip scarecrow is enabled. is that intentional?

Dumbledork01 replied:
Oops, it definitely is. I meant to include the positional logic as an && to it, but I left it out for debugging purposes. I'll add that back in!

Dumbledork01 committed:
Fixed positional Scarecrow logic

briaguya0 commented on soh/src/overlays/actors/ovl_En_Kakasi2/z_en_kakasi2.c:
```diff
@@ -118,9 +118,10 @@ void func_80A90264(EnKakasi2* this, GlobalContext* globalCtx) {
-    if ((BREG(1) != 0) && (this->actor.xzDistToPlayer < this->maxSpawnDistance.x) &&
-        (fabsf(player->actor.world.pos.y - this->actor.world.pos.y) < this->maxSpawnDistance.y)) {
-
+    if ((((BREG(1) != 0) || ((CVar_GetS32("gSkipScarecrow", 0) != 0) && 
+            (globalCtx->msgCtx.ocarinaAction == OCARINA_ACTION_FREE_PLAY) && gSaveContext.scarecrowSpawnSongSet)) &&
+            (this->actor.xzDistToPlayer < this->maxSpawnDistance.x) &&
+            (fabsf(player->actor.world.pos.y - this->actor.world.pos.y) < this->maxSpawnDistance.y))){
```
this turns into
```c
if (
    (
     (
      (BREG(1) != 0) ||
      (
       (CVar_GetS32("gSkipScarecrow", 0) != 0) && 
       (globalCtx->msgCtx.ocarinaAction == OCARINA_ACTION_FREE_PLAY) &&
        gSaveContext.scarecrowSpawnSongSet
      )
     ) &&
    (this->actor.xzDistToPlayer < this->maxSpawnDistance.x) &&
    (fabsf(player->actor.world.pos.y - this->actor.world.pos.y) < this->maxSpawnDistance.y)
   )
  ){
```

you can get rid of a set of parens

Dumbledork01 committed:
Fixed style

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Scarecrow-Song-Skip

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Scarecrow-Song-Skip

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Scarecrow-Song-Skip

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Scarecrow-Song-Skip

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Scarecrow-Song-Skip
